osx-utils/macos-bundler.py

```diff
@@ -0,0 +1,342 @@
+		-Dlicense="{str(performous_source_dir / 'LICENSE.md')}" \
+		"Performous-{package_version}" \
+		"{outFile}"
+	""")
+
+def bundle_libs():
+	global performous_out_dir
+	print("Copying dependencies and fixing linkage inside Performous.app...")
+	
+	execute(fr"""
+		dylibbundler -od -b \
+		-x "{performous_out_dir}/Performous.app/Contents/MacOS/Performous" \
+		-d "{performous_out_dir}/Performous.app/Contents/Resources/lib" \
+		-p @executable_path/../Resources/lib/
+	""")
+	return
+
+usageHelp = f"""\nPerformous macOS Bundler
+
+Usage:
+	macos_bundler.py [--arch <architecture>]... [--prefer-macports | --prefer-homebrew] [options]
+	macos_bundler.py [options]
+
+Options:
+	-h --help  Show this help message.
+	-b --preserve-build  Don't remove temporary build files before starting.
+	-d --debug  Create .app for debugging (skip copying and relinking of dependencies, as well as creation of .dmg package)
+	-f --flat-output  Put output directly in the output folder, without versioned folders.
+	-j <n>, --jobs <n>  Argument passed to make, specifying the max number of jobs to run. Defaults to the output of os.cpu_count() [default: {str(cpu_count())}]
+	-k --no-clean  Don't erase tree structure of the .app before beginning work.
+	-n --no-regenerate  Don't regenerate the CMake build system.
+	-v <version>, --package-version <version>  Sets the version on the Info.plist of the created .app. By default, it's calculated according to the latest stable version and hash of the latest commit.
+	-x --xcode-project  Generate an XCode project suitable for debugging. Note: to use this project, it's necessary to build the install target from XCode and then the performous target (or, alternatively, mirror the settings for the performous scheme on the install target)
+	--script-debug  Print the resolved arguments and options passed to this utility.
+
+Environment:
+	--arch <architecture>...  Target architecture names passed to the compiler. [default: x86_64] 
+	--cc <path/to/compiler>  Change C compiler [default: /usr/bin/clang]
+	--cxx <path/to/compiler>  Change C compiler [default: /usr/bin/clang++]
+	-p <prefix>, --prefix <prefix>  Set prefix path for searching of libraries and headers. By default, the tool tries to detect whether MacPorts or Homebrew are installed and the prefix is set accordingly. Note: If both are installed, you can choose.
+	(--prefer-macports | --prefer-homebrew)  Prefer either MacPorts or Homebrew.
+	-s <path>, --source <path>  Path to the Performous source. Defaults to ../
+	-o <path>, --output <path>  Path where the .app will be built. Defaults to <performous-source>/osx-utils/out[/xcode]
+	-t <target>, --target <target>  macOS target version. Defaults to the currently running version, as reported by platform.mac_ver() [default: {str(float('.'.join(platform.mac_ver()[0].split('.')[:2])))}]"""
+
+if __name__ == "__main__":
+	arguments = docopt(docstring = usageHelp, default_help = False)
```

Comment:
why use docopt over the python-included argparse?

Reply:
Because it gives us a great user experience (my main motivation in doing this) and makes it easier change down the line.

Reply:
Python's argparse is quite bad actually. Needs quite a bit of code and still has severe limitations (requiring monkeypatching its code to get to good parsers, a lot of trickery to get well formatted help). I have no experience with docopt yet but I do plan to try it on my next project that needs a simple parser (thanks @Lord-Kamina).

Reply:
If you do (and want the python version) make sure to use `docopt-ng` because the original is essentially abandoned.